### PR TITLE
Goldilocks inverse system call

### PR DESCRIPTION
**Pull request by lvella**
This can help accelerate Plonky3 verification, as the bottleneck is calculating field element inverse.

### CODE REVIEW

File: riscv/src/large_field/code_gen.rs [hunk 1]
```diff
@@ -443,6 +443,31 @@ fn preamble(field: KnownField, runtime: &Runtime, with_bootloader: bool) -> Stri
         tmp3_col = 1 - XXIsZero
     }
 
+    // ================= ground field arithmetic =================
+
+    // Inverts the Goldilocks field value in memory pointed by register X,
+    // returning the low and high limbs in registers Y and Z, respectively.
```
Comment (leonardoalt) [inline riscv/src/large_field/code_gen.rs:449]: comment is outdated

File: riscv/src/large_field/code_gen.rs [hunk 2]
```diff
@@ -443,6 +443,31 @@ fn preamble(field: KnownField, runtime: &Runtime, with_bootloader: bool) -> Stri
...
+    //
+    // Unsolvable if X points to 0.
```
Comment (leonardoalt) [inline riscv/src/large_field/code_gen.rs:451]: this one too

File: riscv/src/large_field/code_gen.rs [hunk 3]
```diff
@@ -443,6 +443,31 @@ fn preamble(field: KnownField, runtime: &Runtime, with_bootloader: bool) -> Stri
...
+    instr invert_gl X, Y
```
Comment (leonardoalt) [inline riscv/src/large_field/code_gen.rs:452]: should we always have the instruction though? for other things we require the coprocessor to be explicitly "requested".
I think it's fine like this, we should anyway add an optimization pass that removes unused instructions

Comment (lvella) [reply riscv/src/large_field/code_gen.rs:452]: I think this is fine because it is a single instruction, not a whole new machine.

Comment (leonardoalt) [reply riscv/src/large_field/code_gen.rs:452]: that's still at least a couple extra columns if unused, but fine

**Review by leonardoalt — DISMISSED** [review]
Wait, why are changes to the executor not needed?

File: riscv-runtime/src/goldilocks.rs [hunk 4]
```diff
@@ -20,11 +23,22 @@ impl Goldilocks {
         Self(value)
     }
 
+    /// Returns the inverse of the field element.
+    pub fn inverse(self) -> Self {
+        let mut low = self.0 as u32;
+        let mut high = (self.0 >> 32) as u32;
+        unsafe {
```
Comment (Schaeff) [inline riscv-runtime/src/goldilocks.rs:30]: Would it make sense to only have this function on the riscv target?

Comment (leonardoalt) [reply riscv-runtime/src/goldilocks.rs:30]: wdym by target, the Rust rv32 target, or our riscv asm machine?

File: riscv/src/large_field/code_gen.rs (hunk 1 above)
Comment (lvella) [reply riscv/src/large_field/code_gen.rs]: fixed

File: riscv/src/large_field/code_gen.rs (hunk 2 above)
Comment (lvella) [reply riscv/src/large_field/code_gen.rs]: fixed

**Comment by lvella** [comment]
> Wait, why are changes to the executor not needed?

Good question! I forgot about it, but it worked anyway. I have no idea what is happening.

File: riscv-runtime/src/goldilocks.rs (hunk 4 above)
Comment (Schaeff) [reply riscv-runtime/src/goldilocks.rs:30]: My bad, thought this was in number

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**